### PR TITLE
Refine nydus-image: part 1

### PR DESCRIPTION
**Pull request by jiangliu**
Refine nydus-image

### CODE REVIEW

File: src/bin/nydus-image/merge.rs [hunk 1]
```diff
@@ -115,26 +125,35 @@ impl Merger {
...
+                        bail!(
+                            "can not merge bootstraps with inconsistent chunk size, current bootstrap {:?} with chunk size {:x}, expected {:x}",
+                            bootstrap_path,
+                            blob_ctx.chunk_size,
+                            chunk_size.unwrap(),
+                        );
+                    }
                     parent_blob_added = true;
                 }
+                // TODO: this seems wrong, blobs from chunk dict will be added more than one time.
```
Comment (imeoer) [inline src/bin/nydus-image/merge.rs:156]: What's the problem here?

File: rafs/src/metadata/chunk.rs [hunk 2]
```diff
@@ -45,97 +48,127 @@ impl ChunkWrapper {
...
 
+    /// Get size of compressed chunk data.
     pub fn compressed_size(&self) -> u32 {
         match self {
             ChunkWrapper::V5(c) => c.compressed_size,
             ChunkWrapper::V6(c) => c.compressed_size,
         }
     }
 
+    /// Get size of compressed chunk data.
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply rafs/src/metadata/chunk.rs:107]: Fixed

File: rafs/src/metadata/chunk.rs [hunk 3]
```diff
@@ -45,97 +48,127 @@ impl ChunkWrapper {
...
+
+    /// Get chunk index into the RAFS chunk information array, used by RAFS v5.
+    pub fn index(&self) -> u32 {
+        match self {
+            ChunkWrapper::V5(c) => c.index,
+            ChunkWrapper::V6(c) => c.index,
+        }
+    }
+
+    /// Get chunk index into the RAFS chunk information array, used by RAFS v5.
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply rafs/src/metadata/chunk.rs:155]: Fixed

File: src/bin/nydus-image/core/node.rs [hunk 4]
```diff
@@ -695,20 +659,17 @@ impl Node {
...
-
                 for name in entry_names.iter() {
                     dir_data.extend(name.as_bytes());
                 }
 
                 let tail_off = match self.v6_datalayout {
                     EROFS_INODE_FLAT_INLINE => self.v6_offset + self.v6_size_with_xattr() as u64,
                     EROFS_INODE_FLAT_PLAIN => dirent_off,
-                    _ => unimplemented!(),
+                    _ => panic!("unsupported RAFS v6 inode layout for directory"),
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply src/bin/nydus-image/core/node.rs:670]: Fixed

File: src/bin/nydus-image/merge.rs (hunk 1 above)
Comment (jiangliu) [reply src/bin/nydus-image/merge.rs:156]: Fixed

Comment (jiangliu) [reply src/bin/nydus-image/merge.rs:156]: Say we are going to merge two images, and those two image reference a common layer (LayerA), then there will be two LayerA in the merged blob table.

Comment (jiangliu) [reply src/bin/nydus-image/merge.rs:156]: We need something like this
```
diff --git a/src/bin/nydus-image/merge.rs b/src/bin/nydus-image/merge.rs
index 3b25398f2..bf63ebd7d 100644
--- a/src/bin/nydus-image/merge.rs
+++ b/src/bin/nydus-image/merge.rs
@@ -153,9 +153,18 @@ impl Merger {
                     }
                     parent_blob_added = true;
                 }
-                // TODO: this seems wrong, blobs from chunk dict will be added more than one time.
-                blob_idx_map.push(blob_mgr.len() as u32);
-                blob_mgr.add(blob_ctx);
+
+                let mut found = false;
+                for (idx, blob) in blob_mgr.get_blobs().iter().enumerate() {
+                    if blob.blob_id == blob_ctx.blob_id {
+                        blob_idx_map.push(idx as u32);
+                        found = true;
+                    }
+                }
+                if !found {
+                    blob_idx_map.push(blob_mgr.len() as u32);
+                    blob_mgr.add(blob_ctx);
+                }
             }
 
             if let Some(tree) = &mut tree {
```

Comment (imeoer) [reply src/bin/nydus-image/merge.rs]: Okay, good catch!

**Review by imeoer — APPROVED** [review]

**Pull request merged**